### ISSUE TITLE
Volatile nonsense

### ISSUE THREAD

**Issue by LuisAyuso** [insieme/insieme#81]
Recently I crashed into this:
```
let fun000 = fun() -> int<4> {
    decl ref<volatile<int<4>>> v1 = ( var(undefined(type<volatile<int<4>>>)));
    (v1 := volatile_make(10));
    return volatile_read(( *v1));
};

// Inspire Program 
//  Entry Point: 
fun000
```

The current state of the volatile representation is not accurate.
- the Declaration qualifies the Type int as ```volatile``` and declares a variable```ref``` of type``` volatile<int<4>>```.  This means that any read of the value contained inside of var ```v1``` is of type ```volatile<int<4>>``` which is a nonsense, since volatile is an operation that affects only the read/write of a variable, and therefore needs to be applied to ```ref<int<4>>```.
- The ```volatile_make``` function makes little sense as well: it transforms a type ```int<4>``` into ```volatile<int<4>>```  and serves the purpose of making the semantics checks happy, while the write operation (left side of assignment) has no semantic indication of the volatile operation. 

The misfortuned design of this feature can only explain by the need of transmitting the volatile tag to the backend, meanwhile little information about the volatile semantics is preserved.

Solutions:
-------------

- Write a whole new set of operations that attach to ```ref``` volatiles. No ```make_volatile```, instead make a ```read_volatile (volatile<ref<'a>>)->'a``` and ```write_volatile(volatile<ref<'a>>, 'a)``` (which is assignment for volatiles). A third operation might be needed ```decay_to_ref(volatile<ref<'a>>)->ref<'a>``` to pass by reference a volatile into a function which makes no use of volatile in the parameter.
**The Problem:**  Parameters and member fields marked volatile might not have a ref in ir (parameters by value, structs have a single ref for the whole struct) 

- Tag the ref: Since we care pretty little or nothing about volatiles in the analysis. we should choose a fast and robust mechanism to pass this to the backend: Make a RK_Volatile and put it into the ref type.
**The Problem:**  idem as before... 

- Annotate: just tell the backend that there it goes a volatile flag.
**The Problem:**  Annotations get lost... 

What to do?

**Comment by HerbertJordan** [comment]
It was back in good old 2011, it was Christmas time and the big idea was to get finally ready for research.  And to do so, it was decided to run a tone of variations of the NAS benchmarks over the holidays to establish a solid data set for future machine learning activities. Oh yes, good old times ...

Anyway, the only thing between us and the proper support of the NAS benchmarks was the volatile keyword. And we have been young, and nobody had yet a good understanding of the IR (nor a clue in which direction it will develop) so that we just went on to get it from the FE to the BE and be fine with it. And we saw that it was fine. At least nobody ever realized so far that it is completely brocken.

So how to fix it ...

* **Annotations:** not an option. Annotating types / variables => sharing / renaming => big mess
* **abstract type + abstract operators** by introducing a ``volatile<'a>`` type and a cast from and to ``ref<'a>`` as suggested by @LuisAyuso would be a solution, but it will just solve the surve to forward the volatile mark form the FE to the BE, os the current solution does
* **derived type + derived operators** by introducing a type ``struct _volatile { value : ref<'a> }`` and corresponding conversion operations as derived operations we have the same benefits as the abstract type solution, but in addition the analysis will be able to understand that it is actually a reference and data can be transfered through it -- without updating analysis at all. So, FE, BE and Analysis are fine .. almost. **The Problem:** we do not cover the *parallel* semantic of volatile

In the semantic definition of our IR any reference is implicitly shared among all threads. We have two interpretations: 
* option A - sequential consistency. In this case all read/write operations to shared references are atomic; this would however imply that the BE is always producing incorrect code whenever not wrapping any variable into an atomic variable. This is good for theory, but not good for performance nor a practical implementation
* option B: if there is a race condition between threads regarding a specific reference, any read targeting that reference will be undefined -- starting from the last sync point before the code section constituting the race condition until the following race-condition free write; this is the approach taken by C/C++ and OpenMP. And that's the interpretation used by the backend.

 For formal details, see page 95ff in [here](http://www.dps.uibk.ac.at/~csaf7445/pub/phd_thesis_jordan.pdf).

Now, a variable in C/C++ wouldn't be marked volatile if there wouldn't be a race condition on it. So in this case, we simply can not model an equivalent program in our IR following interpretation B.

Unless: we are actually using the construct offered by our IR for inter-thread communication -- channels.
* volatile variables can be modeled using **channels**. As for the solution based on composed types and derived operators, we create a struct containing a reference + a channel of type ``channel<unit,1>`` -- the latter is used like a lock for the reference operation. This way, all accesses are synchronized, and the semantic is properly modelled. 

Does this sound plausible?

**Comment by LuisAyuso** [comment]
:+1:  
As usual, very correct. 

May concerns are a little more pragmatic. Developing a new derived type that makes use of channels seems to model the case. The question is: How long to implement? Does it pay of?

There is a thing you did not commented. Volatiles are (in practice) applied to ref types. In IR we might not have those refs anymore because of nesting (struct with a volatile field) and function call (volatile paramenter). Do the channels rely on a ref somewhere to do the write? i mean, Does the channel need to be dependent on a ref<int> somewhere? lang basic says channel of 'a, but i dont see how they can work if 'a is not a mutable type. excuse me for my ignorace, but channels were always something very away from my scope.

**Comment by HerbertJordan** [comment]
Well, the implementation of a abstract or derived operator is basically the same. You simply just use a different macro to describe it in the language extension.

The problem with the volatile property being passed along with the reference, when accessing a nested structure is, however, a bit more tricky. This would be asking for another type of reference, as we did it with const. But then, we suddenly have a new concept in our parallel model; Also, on a related matter, modeling atomic values has the same problem, with the additional issue of C/C++ support for different [memory orders](http://en.cppreference.com/w/cpp/atomic/memory_order). If it is your interrest to get this all nicely cleaned up, I have to thing about this in more detail.

Is this an urgent issue?

**Comment by LuisAyuso** [comment]
Not urgent, I started a deep cleanup of the frontend visitors, delegating all non core elements (decorators and other details) into extensions. Currently I work on a branch where I have a test case for all the volatile uses I found (I will do so for atomics) .

I agree. Introducing a wrapper type, which holds a ref inside, can lead to a misleading memory model. But the problem only comes out with the volatile. The cpp reference should be correct since there is a ref inside the struct. It is the initialization of the super type the one that allocates a new memory region for the field or uses an existing one somewhere else.

So, to summarize:
We need to reconsider the IR to model the behavior of: 
- [ ] Atomic types
- [ ] volatile types

In: 
- ref<T> 
- struct{ T field }   (the ref appears when allocate the whole struct)
- collection< T > (for array, vector... simd vectors? ) 
- Use as left side: Assign, pass by reference, any other??
- use ar right side: when clang LtoR cast appears. any other??

**Comment by LuisAyuso** [comment]
moved to #83
